### PR TITLE
[15.0][MIG] base_model_restrict_update

### PR DESCRIPTION
**Pull request by Saran440**
No change code

### CODE REVIEW

**Comment by StefanRijnhart** [comment]
Thanks! Can you include https://github.com/OCA/server-tools/pull/2626?

**Comment by Saran440** [comment]
@StefanRijnhart Thanks for review. This PR add commit https://github.com/OCA/server-tools/pull/2626
and 1 commit for fix error

Peek 2023-06-15 15-35

File: base_model_restrict_update/models/res_users.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    )
+
+    @api.constrains("is_readonly_user")
+    def _check_is_readonly_user(self):
+        for user in self:
+            if self.env.ref("base.group_system") in user.groups_id:
+                raise UserError(_("You cannot set admin user as a readonly user."))
+
+    def toggle_unrestrict_model_update(self):
+        pass
```
Comment (StefanRijnhart) [inline base_model_restrict_update/models/res_users.py]: If it is necessary to add this code override to make the GUI widget work (and I played with it a bit, and I don't see another way myself), maybe Odoo is telling us we should not be doing this 'boolean field in a stat button' thing. Would you consider putting the binary toggles somewhere else on the user form?

Comment (yostashiro) [reply base_model_restrict_update/models/res_users.py]: I also checked the behavior and found that if you replace the buttons with the following divs then the presentation and behavior become OK without the dummy toggle methods:

```.xml
                <div
                    name="toggle_unrestrict_model_update"
                    class="btn-group btn oe_stat_button"
                    attrs="{'invisible': [('is_readonly_user', '=', True)]}"
                    style="align-items: center;"
                >
                    <span style="padding: 0 10px;">Unrestrict Update</span>
                    <field name="unrestrict_model_update" widget="boolean_toggle" />
                </div>
                <div
                    name="toggle_is_readonly_user"
                    class="btn-group btn oe_stat_button"
                    style="align-items: center;"
                >
                    <span style="padding: 0 10px;">Read-only</span>
                    <field name="is_readonly_user" widget="boolean_toggle" />
                </div>
```

But I agree with Stefan that we may put the toggles elsewhere if it becomes a maintenance burden to put them in the button box.

Comment (StefanRijnhart) [reply base_model_restrict_update/models/res_users.py]: Oh, I can live with this `div` solution!

Comment (Saran440) [reply base_model_restrict_update/models/res_users.py]: @StefanRijnhart @yostashiro Thanks for solution :) I fixed it.

File: base_model_restrict_update/views/res_users_views.xml [hunk 2]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+
+    <record id="view_users_form" model="ir.ui.view">
+        <field name="name">res.users.form</field>
+        <field name="model">res.users</field>
+        <field name="inherit_id" ref="base.view_users_form" />
+        <field name="arch" type="xml">
+            <div class="oe_button_box" name="button_box" position="inside">
+                <div
```
Comment (yostashiro) [inline base_model_restrict_update/views/res_users_views.xml:10]: A little comment would be nice.

```suggestion
                <!-- "button" tag cannot be used without object/action assignment as of 15.0, therfore we have chosen to use "div" here. -->
                <div
```

**Review by yostashiro — APPROVED** [review]
Functional test.  Thanks for the update!

File: base_model_restrict_update/models/res_users.py (hunk 1 above)
Comment (StefanRijnhart) [reply base_model_restrict_update/models/res_users.py]: Thanks! I think this all counts as migration stuff so can you squash the two last commits into the migration commits?

File: base_model_restrict_update/views/res_users_views.xml (hunk 2 above)
Comment (StefanRijnhart) [reply base_model_restrict_update/views/res_users_views.xml:10]: Not sure. This is the story of the road we traveled, but maybe we can let the code speak for itself here :grinning:?

File: base_model_restrict_update/models/res_users.py (hunk 1 above)
Comment (Saran440) [reply base_model_restrict_update/models/res_users.py]: I think migration commit should original and rebase https://github.com/OCA/server-tools/pull/2626 and [Fix error](https://github.com/OCA/server-tools/pull/2620#discussion_r1232949173) into 1.

**Review by StefanRijnhart — APPROVED** [review]
Thanks for the update!

**Pull request merged**